### DISCUSSION TITLE
Gesture Fail while button click?

### DISCUSSION THREAD

**Discussion by DevTKSS** [unoplatform/uno#18563] [Q&A] [answered]
Newbe question/bug(?) incoming
Can someone review this and tell me what am I doing wrong?

Since WinSDK is out of order at the moment, i am targeting Desktop in visual studio 2022
Same behaviour in xaml and c#-markup
using mvux

terminal tells me: "The specified framework 'Microsoft.WindowsDesktop.App' is not present in the previously loaded runtime.
fail: Microsoft.UI.Input.GestureRecognizer[0]
      Microsoft.UI.Xaml.Controls.ScrollContentPresenter Inconsistent state, we already have a pending gesture for a pointer that is going down. Abort the previous gesture.
fail: Microsoft.UI.Input.GestureRecognizer[0]
      Microsoft.UI.Xaml.Controls.ScrollContentPresenter Inconsistent state, we already have a pending gesture for a pointer that is going down. Abort the previous gesture."

How to reproduce that:
clicked on 1. the dropdown button 2. the first dropdownitem which haves the new command and after no pop up the seconditem... each item fail message...
I have no clue what I should correct, but might be related to the async?

Untill now I did only vb.net with winforms and never done anything async

Thanks in advance for your help!
Kind regards
![Screenshot_2024-10-23_231203](https://github.com/user-attachments/assets/8b37e3c6-37d3-47dc-b614-1641971b4d66)
![Screenshot_2024-10-23_231401](https://github.com/user-attachments/assets/0eb748e4-e918-4f05-b6f3-beee9a7da053)
![Screenshot_2024-10-23_231445](https://github.com/user-attachments/assets/5252f946-0bc1-45ac-a3c8-8d176c06915d)

**Comment by MartinZikmund** [comment, marked as answer]
We are aware of this warning message (issue https://github.com/unoplatform/uno/issues/7646), but it should not affect functionality of the app, does it have any adverse effect?

**Reply by MartinZikmund** [reply]
@DevTKSS Can you please provide the source code as a minimal `.zip` repro so we can use it to debug?

**Reply by DevTKSS** [reply]
Could you define "minimal".zip repo? I just know normal ZIP, but no "minimal"😅😅😅

**Reply by morning4coffe-dev** [reply]
@DevTKSS What Martin meant was minimal code sample that we could reproduce the issue with on our side and just send the project in .zip format.

**Reply by MartinZikmund** [reply]
@DevTKSS Thank you for the repro, let's continue tracking this issue in #7646

**Comment by DevTKSS** [comment]
@MartinZikmund: I think I may have the reason for this behaviour on my application, but I am struggeling with how to do it right. Maybe some more experienced could help with the final solution for that...

I let gpt check the code of both (MainPage, MainModel) and it came back with the answer, that this gesture fail situation will most likly result from the fact that I am calling the task to show the ContentDialog from a MenuFlyoutItem implemented along the winUI in the dropdown button. So the dropdownbutton will collapse after clicking on the item in its children while the result of the contentDialog Task is finished.

I read the documentation pages but somehow I dont find the right way/syntax for it to solve.

Template Button in the StackPanel:

new Button()
    .Content("Go to Second Page")
    .AutomationProperties(automationId: "SecondPageButton")
    .Command(() => vm.GoToSecond),

which called the async Task in the MainModel using the name Property there:

public partial record MainModel
{
    private INavigator _navigator;

    public MainModel(
        IOptions<AppConfig> appInfo,
        INavigator navigator)
    {
        _navigator = navigator;
        Title = "Main";
        Title += $" - {appInfo?.Value?.Environment}";
    }

    public IState<string> Name => State<string>.Value(this, () => string.Empty);

    public async Task GoToSecond()
    {
        var name = await Name;
        await _navigator.NavigateViewModelAsync<SecondModel>(this, data: new Entity(name!));
    }

Okay, so since I would like to do something like this [winui3gallery://item/ContentDialog](url) mentioned in the WinUI Gallery to the ContentDialog, where it uses just a button to show a dialog and put in the result into the textblock right from the button (okay, my textblock is below, but should not really matter where it is)

new Button()
    .Content("Go to Second Page")
    .AutomationProperties(automationId: "SecondPageButton")
    .Command(() => vm.GoToSecond),
new DropDownButton() /// here the added content starts, resource of the structure: [winui3gallery://item/DropDownButton](url) from xaml to c# markup
    .Name("btnDropThatDown")
    .Content("Let's drop that down!")
    .Flyout(
    new MenuFlyout()
     .Placement(FlyoutPlacementMode.Bottom)
     .Items(
         new MenuFlyoutItem()
             .Text("Let's go right!")
             .Icon(new FontIcon().Glyph("\uE72A"))
             .AutomationProperties(automationId: "Let's go right!")
             .Command(() => vm.ShowDialogAsync(page.XamlRoot,
                                               "You dropped it down and gone right!",
                                               "Buh Ja!",
                                               "Let's go back!",
                                               "Want to stay!",
                                               "WTF? Go away!")
             ),
             new MenuFlyoutItem()
             .Text("Let's go left!")
             .Icon(new FontIcon().Glyph("\uE72B")),
         new MenuFlyoutItem()
             .Text("Let's turn around!")
             .Icon(new FontIcon().Glyph("\uE72C"))
 )),

What I also added was the input parameters for the Dialog, to maybe make this more... flexible? I like to follow known patterns, which means: In WindowsForms there is the overloadable Method "DialogResult Messagebox.Show(Win32Window WindowOwner, String text, String title, [MessageBoxButton button = MessageBoxButton.OK], [MessageBoxIcon(Enum) icon = null], [button defaultButton = FirstButton])" which in my opinion is really nice because there is just one Method to call for showing this MessageBox and getting the Options to add the properties just there in the call. I would love to generate a template, service... please excuse, I just  dont know the correct area for this task, I am open for correction to know this in the future!
I never worked with Task in the past, but seems like a ContentDialog wants to be placed in a async Method.
So doing this "ShowDialogAsync" with delivering the desired text values and if possible the text for content and buttons (dont know how to deliver the needed xamlroot also since it seems to be required to be the MainPage instance isnt it? In my thinking this xamlRoot seems to be the synonym for the Win32Window I know from WinForms... But this tasks seems to also want this cancellationToken thing which I do not really understand since there is a cancelButton?

Okay, so since in the template uses a IState<string> kind of... Property? which is filled from the TextBox, I had this Idea to do my result just  other way around:

    public IState<string> DialogResult => State<string>.Value(this, () => string.Empty);

public async Task ShowDialogAsync(XamlRoot xamlRoot,
                                  string title,
                                  string message,
                                  string primaryButtonText,
                                  string? secondaryButtonText = null,
                                  string? closeButtonText = null)
{
    ContentDialog dialog = new ContentDialog
    {
        Title = title,
        Content = message,
        PrimaryButtonText = primaryButtonText,
        SecondaryButtonText = secondaryButtonText,
        CloseButtonText = closeButtonText,
        XamlRoot = xamlRoot,
    };
    if (secondaryButtonText is not null)
    {
        dialog.SecondaryButtonText = secondaryButtonText;
    }
    if (closeButtonText is not null)
    {
        dialog.CloseButtonText = closeButtonText;
    }
    ContentDialogResult result = await dialog.ShowAsync();

    await DialogResult.SetAsync(result switch
    {
        ContentDialogResult.Primary => primaryButtonText,
        ContentDialogResult.Secondary => secondaryButtonText ?? "secondaryButton",
        _ => closeButtonText
    });
}
So what I am trying to to is, after calling the task to show my dialog, it should write the result in the property DialogResult from the MainModel which uses IState to be an Observable. Maybe a IFeed would be also possible if it would be able to bind the DialogResult directly to this Task... but don't know if that would work some way.

So in my thought, this click would call the task to run with lambda and do not really need to watch that anymore, so should not any longer do this gesture fail in my opinion.
To get the result in my UI back again, I try to use a Binding in the TextBlock below the DropDownButton and also bind the Visibility Property of that to the text of the TextBlock itself:

 new TextBlock()
     .AutomationProperties(automationId: "DialogResultOutput")
     .Text(x => x.Binding(()=>vm.DialogResult))
     .Visibility(x => x.RelativeSource<TextBlock>(RelativeSourceMode.Self)
                       .Binding(t=> string.IsNullOrEmpty(t.Text) ? Visibility.Collapsed : Visibility.Visible))

but sadly I seem to do something wrong because now, the compiler tells me the /unsave thing, other than before and still do not even show the dialog... maybe the gesture fail is happening because it opened the dialog (whereever) and wants me to answer it
Can someone point out my fault?

**Reply by DevTKSS** [reply]
zip file was too big to attach, uploaded it in my github as a new repo, so you can review and get it there: https://github.com/DevTKSS/MyUnoApp.git
